File: app/src/main/java/vng/zalo/tdtai/zalo/zalo/views/home/fragments/chat_fragment/room_activity/RoomActivity.java
package vng.zalo.tdtai.zalo.zalo.views.home.fragments.chat_fragment.room_activity;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textfield.TextInputEditText;

import java.util.List;

import javax.inject.Inject;

import vng.zalo.tdtai.zalo.R;
import vng.zalo.tdtai.zalo.zalo.dependency_factories.chat_activity.DaggerRoomActivityComponent;
import vng.zalo.tdtai.zalo.zalo.dependency_factories.chat_activity.RoomActivityModule;
import vng.zalo.tdtai.zalo.zalo.models.Message;
import vng.zalo.tdtai.zalo.zalo.utils.Constants;
import vng.zalo.tdtai.zalo.zalo.utils.MessageDiffCallback;
import vng.zalo.tdtai.zalo.zalo.viewmodels.RoomActivityViewModel;

public class RoomActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = RoomActivity.class.getSimpleName();

    private RecyclerView recyclerView;
    private TextInputEditText msgTextInputEditText;
    private ImageView uploadFileImgButton;
    private ImageView voiceImgButton;
    private ImageView pictureImgButton;
    private ImageView sendMsgImgButton;

    @Inject
    RoomActivityViewModel viewModel;

    RoomActivityAdapter adapter;
    LinearLayoutManager layoutManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        DaggerRoomActivityComponent.builder()
                .roomActivityModule(new RoomActivityModule(this))
                .build().inject(this);

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_room);

        Toolbar toolbar = findViewById(R.id.toolbarRoomActivity);
        toolbar.setTitle(getIntent().getStringExtra(Constants.ROOM_NAME));
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        } else {
            Log.e(TAG, "actionBar is null");
        }

        recyclerView = findViewById(R.id.recyclerViewRoomActivity);

        layoutManager = new LinearLayoutManager(this);

        recyclerView.setLayoutManager(layoutManager);

        viewModel.liveMessages.observe(this, new Observer<List<Message>>() {
            @Override
            public void onChanged(List<Message> messageList) {
                adapter.submitList(messageList, new Runnable() {
                    @Override
                    public void run() {
                        scrollRecyclerViewToLastPosition();
                    }
                });
                Log.d(TAG, "onChanged liveData");
            }
        });

        adapter = new RoomActivityAdapter(this, new MessageDiffCallback());

        recyclerView.setAdapter(adapter);

        pictureImgButton = findViewById(R.id.pictureImgButton);
        sendMsgImgButton = findViewById(R.id.sendMsgImgButton);
        uploadFileImgButton = findViewById(R.id.uploadFileImgButton);
        voiceImgButton = findViewById(R.id.voiceImgButton);

        msgTextInputEditText = findViewById(R.id.msgTextInputEditText);
        msgTextInputEditText.addTextChangedListener(new InputTextListener());

        sendMsgImgButton.setOnClickListener(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_room_activity, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.sendMsgImgButton:
                String textToSend = msgTextInputEditText.getText() == null ? "" : msgTextInputEditText.getText().toString();
                viewModel.addMessageToFirestore(textToSend);

                msgTextInputEditText.setText("");
                break;
        }
    }

    class InputTextListener implements TextWatcher {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            if (s.length() == 0) {
                sendMsgImgButton.setVisibility(View.GONE);
                uploadFileImgButton.setVisibility(View.VISIBLE);
                voiceImgButton.setVisibility(View.VISIBLE);
                pictureImgButton.setVisibility(View.VISIBLE);
            } else {
                sendMsgImgButton.setVisibility(View.VISIBLE);
                uploadFileImgButton.setVisibility(View.GONE);
                voiceImgButton.setVisibility(View.GONE);
                pictureImgButton.setVisibility(View.INVISIBLE);
            }
        }
    }

    void scrollRecyclerViewToLastPosition() {
        recyclerView.scrollToPosition(Math.max(0, adapter.getItemCount() - 1));
    }
}

File: app/src/main/java/vng/zalo/tdtai/zalo/zalo/views/home/fragments/chat_fragment/room_activity/RoomActivity.kt
package vng.zalo.tdtai.zalo.zalo.views.home.fragments.chat_fragment.room_activity

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.activity_room.*
import vng.zalo.tdtai.zalo.R
import vng.zalo.tdtai.zalo.zalo.dependency_factories.chat_activity.DaggerRoomActivityComponent
import vng.zalo.tdtai.zalo.zalo.dependency_factories.chat_activity.RoomActivityModule
import vng.zalo.tdtai.zalo.zalo.utils.Constants
import vng.zalo.tdtai.zalo.zalo.utils.Utils
import vng.zalo.tdtai.zalo.zalo.viewmodels.RoomActivityViewModel
import javax.inject.Inject
import kotlin.math.max

class RoomActivity : AppCompatActivity(), View.OnClickListener {
    @Inject lateinit var viewModel: RoomActivityViewModel
    private lateinit var adapter: RoomActivityAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        DaggerRoomActivityComponent.builder()
                .roomActivityModule(RoomActivityModule(this))
                .build().inject(this)

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_room)

        initView()

        viewModel.liveMessages.observe(this, Observer{ messageList ->
            adapter.messages = messageList
            adapter.notifyDataSetChanged()
            this.scrollRecyclerViewToLastPosition()
//            adapter.submitList(messageList) { this.scrollRecyclerViewToLastPosition() }
            Log.d(Utils.getTag(object {}), "onChanged liveData")
        })
    }

    private fun initView(){
        toolbar.title = intent.getStringExtra(Constants.ROOM_NAME)
        setSupportActionBar(toolbar)

        supportActionBar?.setDisplayHomeAsUpEnabled(true) ?: Log.e(Utils.getTag(object {}), "actionBar is null")

        msgEditText.apply{
            addTextChangedListener(InputTextListener())
            setOnEditorActionListener { _, actionId, _ ->
                when (actionId) {
                    EditorInfo.IME_ACTION_SEND -> {
                        sendMessage()
                        true
                    }
                    else -> false
                }
            }
        }

        sendMsgImgView.setOnClickListener(this)

        adapter = RoomActivityAdapter()
        with(recyclerView){
            adapter = this@RoomActivity.adapter
            layoutManager = LinearLayoutManager(this@RoomActivity)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_room_activity, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when(item.itemId){
            R.id.action_settings -> true
            android.R.id.home -> {
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.sendMsgImgView -> sendMessage()
        }
    }

    private fun sendMessage() {
        val messageContent = msgEditText.text.toString()
        viewModel.addNewMessageToFirestore(messageContent)

        msgEditText.setText("")
    }

    internal inner class InputTextListener : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

        }

        override fun afterTextChanged(s: Editable) {
            if (s.isEmpty()) {
                sendMsgImgView.visibility = View.GONE
                uploadFileImgButton.visibility = View.VISIBLE
                voiceImgButton.visibility = View.VISIBLE
                pictureImgButton.visibility = View.VISIBLE
            } else {
                sendMsgImgView.visibility = View.VISIBLE
                uploadFileImgButton.visibility = View.GONE
                voiceImgButton.visibility = View.GONE
                pictureImgButton.visibility = View.INVISIBLE
            }
        }
    }

    private fun scrollRecyclerViewToLastPosition() {
        recyclerView.scrollToPosition(max(0, adapter.itemCount - 1))
    }
}